src/relay/backend/te_compiler.cc

```diff
@@ -408,27 +462,135 @@ class LowerTensorExpr : public ExprMutator {
     }
 
     tir_call_attrs->metadata.Set("relay_attrs", func->attrs);
+    tir_call_attrs->metadata.Set("all_prim_fn_vars", all_prim_fn_vars);
+
+    if (IsDynamic(func->ret_type)) {
+      // Also lower the dynamic shape function.
+      // Shape function keys use the underlying primitive function as their 'function',
+      // but the generic 'cpu' target as the target since all shape functions run
+      // on the host cpu irrespective of where the primitive runs.
+      // TODO(mbs): Cleanup target handling.
+      Target shape_target("llvm");
+      CCacheKey shape_key(func, shape_target);
+      CachedFunc lowered_shape_func = compiler_->LowerShapeFunc(shape_key);
+      // Capture the shape function's global var and parameters 'states' in call
+      // annotations so calling convention can be recovered.
+      // TODO(mbs): Capture all this as part of a 'call into TIR' construct once available.
+      // The way the shape function calling convention is derived and passed to call sites
+      // via the 'parameter states' could be improved.
+      tir_call_attrs->metadata.Set("prim_shape_fn_var", lowered_shape_func->prim_fn_var);
```

Comment:
Possible follow up, maybe we can refactor into composite structure in the Attrs?

Reply:
Yeah, I think this should be a struct inside the 'call-tir' op.